Handle sequence lengths correctly when exporting RNNs to ONNX

anderspapitto opened the pull request:
PackedSequence: store batch_sizes as tensor rather 
than converting to a list of python integers. This maintains
the invariant that module's inputs/outputs are collections of
Variables.

In particular, this causes the JIT to no longer choke when flattening
and unflattening arguments.

--------------

- when uniform sequence lengths are provided, correctly omit the
  argument when constructing the ONNX graph, so as to not fix the
  graph to the batch size.

- handle PackedSequences by floating them through the graph and
  eliminating them in an optimization pass. ONNX does not have packed
  sequences, but operates on a representation equivalent to
  PaddedSequence, so we hide the representation-switching from ONNX

- as a preliminary step towards handling PackedSequences, not directly
  tied to ONNX export, change batch_sizes from being an argument to
  the RNN operators into being an argument to the forward() function
  of those RNN operators. This more closely models the reality that
  batch_sizes are effectively part of the input sequences.

houseroad commented on torch/csrc/jit/export.cpp:
```diff
@@ -195,12 +195,19 @@ void encodeGraph(onnx::GraphProto * p_g, const std::shared_ptr<Graph> & g, const
     for(auto input : node->inputs()) {
-      p_n->add_input(value_name(input));
+      if (input->node()->kind() == stringToSymbol("OnnxElidedInput")) {
+        p_n->add_input("");
```
Are we going to export input with empty name? We should have a better story about the optional parameter lists.

houseroad replied:
Just talked to other developers, empty string is the current solution for the optional parameter list. So I am not gonna block this PR for this.

houseroad reviewed (commented):
We should add some docs about OnnxElidedInput type node. We should also have a better story for multiple optional parameters.

houseroad commented on torch/csrc/jit/export.cpp:
```diff
@@ -195,12 +195,19 @@ void encodeGraph(onnx::GraphProto * p_g, const std::shared_ptr<Graph> & g, const
       // to help symbolics do the right thing.
       continue;
     }
+    if (node->kind() == stringToSymbol("OnnxElidedInput")) {
```
Shall we have internal string named "kONNXElidedInput"

ezyang commented:
@anderspapitto and I discussed using Undefined instead of introducing a new ElidedOnnxNode.

Commit:
PackedSequence: store batch_sizes as tensor

rather than converting to a list of python integers. This maintains
the invariant that module's inputs/outputs are collections of
Variables.

In particular, this causes the JIT to no longer choke when flattening
and unflattening arguments.

ezyang commented on torch/csrc/jit/export.cpp:
```diff
@@ -190,17 +190,24 @@ void encodeGraph(onnx::GraphProto * p_g, const std::shared_ptr<Graph> & g, const
   for (auto node : g->nodes()) {
-    if (node->kind() == kUndefined && !node->hasUses()) {
+    if (node->kind() == kUndefined) {
       // Undefined nodes never show up in ONNX; they're just a tool
```
This comment is out of date now, right?

anderspapitto replied:
yes, will refine it

ezyang commented on torch/csrc/jit/export.cpp:
```diff
@@ -190,17 +190,24 @@ void encodeGraph(onnx::GraphProto * p_g, const std::shared_ptr<Graph> & g, const
       // Undefined nodes never show up in ONNX; they're just a tool
       // to help symbolics do the right thing.
+      //
+      // One particular use is to implement optional arugments - see
```
arguments

anderspapitto replied:
++

ezyang commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -171,6 +171,34 @@ void fuseTransposeIntoGemm(std::shared_ptr<Graph>& graph) {
+    auto* n = *it;
+
+    // for any PadPacked, if we can trace the PackedSequence back to
+    // an originating PackPadded, elimintate it.
```
eliminate

anderspapitto replied:
++

ezyang commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -171,6 +171,34 @@ void fuseTransposeIntoGemm(std::shared_ptr<Graph>& graph) {
+  for (auto it = graph->begin(); it != graph->end(); ++it) {
+    auto* n = *it;
+
+    // for any PadPacked, if we can trace the PackedSequence back to
```
I'm not sure this transformation is sound. Can you give an example trace where it's necessary to walk across inputs to successfully eliminate the packing?

anderspapitto replied:
well, the whole point of the packed sequence is that there's an RNN in the middle. So it will be (Pack -> RNN -> Unpack). I'll add an example.

anderspapitto replied:
Here's an example export of 3-layer Elman RNN, using packed sequences. (try your best to ignore all the concats and slices cluttering it up, they aren't relevant here)

WIth peephole optimizations turned off, we get

```
graph(%0 : Float(5, 2, 7)
      %1 : Float(3, 2, 13)
      %2 : Long(2)
      %3 : Float(13, 7)
      %4 : Float(13, 13)
      %5 : Float(13)
      %6 : Float(13)
      %7 : Float(13, 13)
      %8 : Float(13, 13)
      %9 : Float(13)
      %10 : Float(13)
      %11 : Float(13, 13)
      %12 : Float(13, 13)
      %13 : Float(13)
      %14 : Float(13)) {
  %15 : Float(10, 7), %16 : Long(5) = PackPadded(%0, %2), scope: RnnModelWithPackedSequence
  %17 : UNKNOWN_TYPE = Constant[value= 10  10  10  10  10  10  10 [ CPUIntTensor{7} ]](), scope: RnnModelWithPackedSequence/RNN[model]
  %18 : UNKNOWN_TYPE = Concat[axis=0](%5, %6), scope: RnnModelWithPackedSequence/RNN[model]
  %19 : UNKNOWN_TYPE = Slice[axes=[0], ends=[1], starts=[0]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %20 : UNKNOWN_TYPE, %21 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%15, %3, %4, %18, %17, %19), scope: RnnModelWithPackedSequence/RNN[model]
  %22 : UNKNOWN_TYPE = Concat[axis=0](%9, %10), scope: RnnModelWithPackedSequence/RNN[model]
  %23 : UNKNOWN_TYPE = Slice[axes=[0], ends=[2], starts=[1]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %24 : UNKNOWN_TYPE, %25 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%20, %7, %8, %22, %17, %23), scope: RnnModelWithPackedSequence/RNN[model]
  %26 : UNKNOWN_TYPE = Concat[axis=0](%13, %14), scope: RnnModelWithPackedSequence/RNN[model]
  %27 : UNKNOWN_TYPE = Slice[axes=[0], ends=[3], starts=[2]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %28 : Float(10, 13), %29 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%24, %11, %12, %26, %17, %27), scope: RnnModelWithPackedSequence/RNN[model]
  %30 : Float(5, 2, 13), %31 : Long(2) = PadPacked(%28, %16), scope: RnnModelWithPackedSequence
  return (%30);
}
```

with peephole optimizations turned back on, we get

```
graph(%0 : Float(5, 2, 7)
      %1 : Float(3, 2, 13)
      %2 : Long(2)
      %3 : Float(13, 7)
      %4 : Float(13, 13)
      %5 : Float(13)
      %6 : Float(13)
      %7 : Float(13, 13)
      %8 : Float(13, 13)
      %9 : Float(13)
      %10 : Float(13)
      %11 : Float(13, 13)
      %12 : Float(13, 13)
      %13 : Float(13)
      %14 : Float(13)) {
  %15 : UNKNOWN_TYPE = Constant[value= 10  10  10  10  10  10  10 [ CPUIntTensor{7} ]](), scope: RnnModelWithPackedSequence/RNN[model]
  %16 : UNKNOWN_TYPE = Concat[axis=0](%5, %6), scope: RnnModelWithPackedSequence/RNN[model]
  %17 : UNKNOWN_TYPE = Slice[axes=[0], ends=[1], starts=[0]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %18 : UNKNOWN_TYPE, %19 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%0, %3, %4, %16, %15, %17), scope: RnnModelWithPackedSequence/RNN[model]
  %20 : UNKNOWN_TYPE = Concat[axis=0](%9, %10), scope: RnnModelWithPackedSequence/RNN[model]
  %21 : UNKNOWN_TYPE = Slice[axes=[0], ends=[2], starts=[1]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %22 : UNKNOWN_TYPE, %23 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%18, %7, %8, %20, %15, %21), scope: RnnModelWithPackedSequence/RNN[model]
  %24 : UNKNOWN_TYPE = Concat[axis=0](%13, %14), scope: RnnModelWithPackedSequence/RNN[model]
  %25 : UNKNOWN_TYPE = Slice[axes=[0], ends=[3], starts=[2]](%1), scope: RnnModelWithPackedSequence/RNN[model]
  %26 : Float(10, 13), %27 : UNKNOWN_TYPE = RNN[activations=[tanh], hidden_size=13](%22, %11, %12, %24, %15, %25), scope: RnnModelWithPackedSequence/RNN[model]
  return (%26);
}
```

ezyang replied:
So, what it looks like to me, is that your optimization is validated by the following equations (in a simplified setting where these ops all have one argument/output; in reality the side conditions are slightly more complicated):

```
PadPacked(RNN(x)) == RNN(PackPadded(x))
PadPacked(PackPadded(x)) = Id
```

So it seems more sound to me to structure the optimization pass this way:

1. Traverse the graph looking for occurrences of `PadPacked`
2. If the output of `PadPacked` is used by the first argument of `RNN`, "push" it past the RNN, turning into a `PackPadded` on the first output of the RNN
3. If the use of a `PadPacked` is a `PackPadded`, eliminate the pair of operations.

There's some leeway in whether or not you should require these to be single-use or not before the optimization fires, in reality they are probably all single use.

How does this sound?

anderspapitto replied:
That sounds pretty good, I'll try out that approach

ezyang commented on torch/nn/modules/module.py:
```diff
@@ -329,7 +329,7 @@ def _tracing_name(self, tracing_state):
 
     def _slow_forward(self, *input, **kwargs):
-        input_vars = tuple(torch.autograd.function._iter_variables(input))
+        input_vars = tuple(torch.autograd.function._iter_variables_permissive(input))
```
Good to have a comment explaining why this one should be permissive.

anderspapitto replied:
me too. I will try to rediscover it

anderspapitto replied:
apparently it doesn't need to be permissive. I'll change it back

ezyang commented on torch/onnx/symbolic.py:
```diff
@@ -650,6 +640,8 @@ def symbolic(g, input, all_weights, h0, **fkwargs):
 
             h_in = h0 if num_layers == 1 else g.op('Slice', h0, axes_i=[0], starts_i=[i], ends_i=[i + 1])
 
+            sequence_lens = g.op("Undefined")
```
Dead code?

ezyang commented on torch/nn/utils/rnn.py:
```diff
@@ -158,7 +180,20 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0):
+
+
+def _symbolic_pad_packed_sequence(g, input, batch_first=False, padding_value=0.0):
+    return g.op("PadPacked", input.data, input.batch_sizes_variable, outputs=2)
```
This isn't a real ONNX op, right? Better say so.

anderspapitto replied:
well, we could make it a real onnx op (if we didn't want to rely 100% on the optimization pass cancelling them out), but for now ++ I'll add a comment

ezyang commented on torch/nn/utils/rnn.py:
```diff
@@ -130,10 +150,12 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0):
-        Tuple of Variable containing the padded sequence, and a list of lengths
-        of each sequence in the batch.
+        Tuple of Variable containing the padded sequence, and Variable
+        containing the list of lengths of each sequence in the batch.
```
I don't think we want to make this BC-breaking change. CC @apaszke what do you think?

apaszke replied:
I’d rather avoid it unless it’s 100% necessary. Is packed sequence support so important? Does ONNX even support variable length sequences?

anderspapitto replied:
ONNX does support variable length sequences (though it's done differently, with a separate, optional `sequence_lens` input).

I believe it is 100% necessary if we want to support variable-length sequences. pytorch itself has some invariants that model inputs/outputs must be (lists of, tuples of) Variables, and PackedSequence violates that. None of the jit/tracing/symbolic machinery works with the current representation.

I am open to investigating alternatives, if anyone has suggestions/ideas of how else this could be made to work.

apaszke commented on torch/nn/_functions/packing.py:
```diff
@@ -5,6 +5,8 @@
 class PackPadded(Function):
     @staticmethod
     def forward(ctx, input, lengths, batch_first):
+        lengths = list(lengths)
```
Is there a reason why we have to do this? It should work just fine with a Variable, right?

anderspapitto replied:
there are some minor things, like calling reversed(Variable) doesn't work. I'll go through all these and double check if I can simplify them.

jekbradbury replied:
I think that's the only issue (but of course fast reversal / negative strides are a longstanding gap in TH)

apaszke replied:
Can we replace `reversed` with a bit more verbose iteration over indices backward, and slice the var at every step.

apaszke commented on torch/nn/_functions/rnn.py:
```diff
@@ -133,6 +133,8 @@ def fac(inner, reverse=False):
 
 
 def VariableRecurrent(batch_sizes, inner):
+    batch_sizes = list(batch_sizes.data)
```
Why does it need to be a list?

apaszke commented on torch/nn/_functions/rnn.py:
```diff
@@ -172,6 +174,8 @@ def forward(input, hidden, weight):
 
 
 def VariableRecurrentReverse(batch_sizes, inner):
+    batch_sizes = list(batch_sizes.data)
```
Same question here

apaszke commented on torch/nn/utils/rnn.py:
```diff
@@ -25,62 +25,69 @@ class PackedSequence(PackedSequence_):
         data (Variable): Variable containing packed sequence
-        batch_sizes (list[int]): list of integers holding information about
-            the batch size at each sequence step
+        batch_sizes_variable (Variable): list of integers holding
```
no longer a list

apaszke commented on torch/nn/utils/rnn.py:
```diff
@@ -25,62 +25,69 @@ class PackedSequence(PackedSequence_):
-            the batch size at each sequence step
+        batch_sizes_variable (Variable): list of integers holding
+            information about the batch size at each sequence step
+        batch_sizes (list of int): equivalent to
```
Eh I think I'm ok with making this always a tensor. This way gives us backward compatibility, but:
- `.batch_sizes` access suddenly becomes expensive
- we never want people to use `.batch_sizes_variable` because it's very verbose

anderspapitto replied:
ok, I will do the simplification

apaszke commented on torch/nn/utils/rnn.py:
```diff
@@ -130,10 +153,12 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0):
     """
-    var_data, batch_sizes = sequence
+    var_data, batch_sizes_variable = sequence
+    batch_sizes = list(batch_sizes_variable.data)
```
Does it need to be a list?

anderspapitto committed:
Handle sequence lengths correctly when exporting RNNs to ONNX

- when uniform sequence lengths are provided, correctly omit the
  argument when constructing the ONNX graph, so as to not fix the
  graph to the batch size.

- handle PackedSequences by floating them through the graph and
  eliminating them in an optimization pass. ONNX does not have packed
  sequences, but operates on a representation equivalent to
  PaddedSequence, so we hide the representation-switching from ONNX

- as a preliminary step towards handling PackedSequences, not directly
  tied to ONNX export, change batch_sizes from being an argument to
  the RNN operators into being an argument to the forward() function
  of those RNN operators. This more closely models the reality that
  batch_sizes are effectively part of the input sequences.

anderspapitto commented:
[WIP] btw. There's one or two other components that I have to do to get this to export correctly

anderspapitto commented:
I switched to the optimization pass strategy that Edward suggested.

Tests are here https://github.com/onnxbot/onnx-fb-universe/pull/466

ezyang commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -167,6 +172,66 @@ void fuseTransposeIntoGemm(std::shared_ptr<Graph>& graph) {
   }
 }
 
+void pushPackingPastRnn(std::shared_ptr<Graph>& graph) {
```
Want a comment here. The discussion from the PR is a good start for the text here.

anderspapitto replied:
++

ezyang commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -167,6 +172,66 @@ void fuseTransposeIntoGemm(std::shared_ptr<Graph>& graph) {
+      continue;
+    }
+    if (n->outputs()[0]->uses().size() != 1) {
+      // For now, only handle the case where there is one consumer.
```
👍 Multiple consumer case probably could be implemented by "inlining" the PackPadded but we definitely don't have to do it until it's needed.

anderspapitto replied:
++ yeah I don't want to feature creep this diff any more than necessary

ezyang commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -167,6 +172,66 @@ void fuseTransposeIntoGemm(std::shared_ptr<Graph>& graph) {
+    }
+
+    // remove PackPadded from in front of the RNN
+    n->outputs()[0]->replaceFirstUseWith(n->inputs()[0]);
```
Minor style point (you don't have to fix): you can access a particular input with `n->input(0)` and `n->output(0)`.

anderspapitto replied:
++

ezyang replied:
If there's only one use of `n->outputs()[0]`, you didn't need to add a `replaceFirstUseWith` function, did you?

I can't think of many situations where you'd actually want `replaceFirstUseWith`.

anderspapitto replied:
(also answering the next comment here)

there is only one use of `n->outputs()[0]` (enforced by line 198), so yes I will switch this to replaceAllUsesWith.

there are actually multiple consumers of `n->outputs()[1]`, and this is fine. The same blob feeds into each level of the RNN as the sequence_lens input (see the code `sequence_lens = ...` in symbolic.py).

Note that pytorch RNNs do a weird thing - at each level, they both consume and return the lengths/batch_sizes (as the second element of the PackedSequence tuple), even though the returned lengths must be the same as the input lengths. ONNX RNNs don't do that weird thing, they only consume the lengths. Therefore I can't mimic the pytorch weirdness exactly, and instead must share the lengths blob between layers

ezyang commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -167,6 +172,66 @@ void fuseTransposeIntoGemm(std::shared_ptr<Graph>& graph) {
+    if (input->kind() != kPackPadded) {
+      continue;
+    }
+    if (input != n->inputs()[1]->node()) {
```
Mmmm... this is not exactly the right way to do this test (although it is unlikely to matter in practice).

What you have said here is that *any* output of `n` is acceptable as the 1-index `input`. That means if I, for example, say:

```
%0, %1 = PackPadded
%2 = PadPacked %1, %0
```

You would accept this as a no-op and optimize it away. But what you really wanted to say is that the first input is the first output of `PackPadded`, and the second input is the second output of `PackPadded`.

anderspapitto replied:
ok, i'll take another look

ezyang commented on torch/nn/_functions/packing.py:
```diff
@@ -14,15 +14,14 @@ def forward(ctx, input, lengths, batch_first):
 
         prev_l = 0
-        for i, l in enumerate(lengths_iter):
+        for i, l in enumerate(reversed(list(lengths))):
```
Was there are a particular reason `lengths_iter` got inlined?

anderspapitto replied:
an attempted clean-up that I ended up changing almost-back. I'll change it back all the way

ezyang commented on torch/nn/modules/rnn.py:
```diff
@@ -142,7 +142,7 @@ def forward(self, input, hx=None):
         if is_packed:
             input, batch_sizes = input
-            max_batch_size = batch_sizes[0]
+            max_batch_size = batch_sizes.data[0]
```
@colesbury Is this going to be affected by the Tensor/Variable merge?

ezyang replied:
OK, it is going to be affected.

To make this code forwards compatible, you'll need to replace `batch_sizes.data[0]` with `int(batch_sizes[0])`.

Back story: in the old days, indexing into 1D tensors would give you a Python number back, but indexing into 1D variables would give you a (scalar) Variable. Since we are merging Variable and Tensor, we are changing the semantics of Tensor indexing to match Variable. This means that the code you wrote above will give you a PyTorch Scalar and will subsequently break when you try to use it later. Using an explicit conversion with `int()` solves the problem, and also solves you having to look at `data`.

anderspapitto replied:
++

ezyang commented on torch/nn/utils/rnn.py:
```diff
@@ -109,10 +110,26 @@ def pack_padded_sequence(input, lengths, batch_first=False):
+
+def _symbolic_pack_padded_sequence(g, input, lengths, batch_first=False):
+    # There currently is no PackPadded operator in ONNX. We rely on an
+    # optimizatino pass to remove this later. It is an error if all
```
optimization

anderspapitto replied:
++

ezyang commented:
This has my STAMP OF APPROVAL.

Flagging @soumith: this as a BC-breaking PR; it changes the batch sizes of `PackedSequence` to be a `Variable`, where it was a list before.

anderspapitto commented:
I believe the onnx-fb-universe tests are still failing due to a deficiency in the test infra - namely, they continue running again the version of caffe2 without the fix/feature that I pushed earlier today.

I'm confirming this with @bddppq offline, and if so that means we can ignore that failure (I've confirmed that the tests that it claims are failing do pass locally)

dzhulgakov commented on torch/autograd/function.py:
```diff
@@ -288,15 +288,22 @@ def _iter(obj):
+            else:
+                res_e, input = unflatten_helper(input, e)
+                res.append(res_e)
+        if isinstance(proto, PackedSequence):
```
this is not a problem of this diff, but I wonder if there's a more elegant solution to handle this than importing random tuples.

One semi-hacky option is to spot any `namedtuple` instances. All of them seem to have `_make` method that takes iterable. So you can do `if hasattr(type(proto), '_make') return type(proto)(*res), input`

Then it'd work for any namedtuple and you'd avoid dependencies

dzhulgakov commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -167,6 +172,85 @@ void fuseTransposeIntoGemm(std::shared_ptr<Graph>& graph) {
+    n->outputs()[1]->replaceAllUsesWith(input->inputs()[1]);
+
+    n->removeAllInputs();
+    input->destroy();
```
theoretically there can be other usages. I guess it's better to keep it and then run DCE pass

anderspapitto replied:
++

dzhulgakov commented on torch/nn/utils/rnn.py:
```diff
@@ -109,10 +110,26 @@ def pack_padded_sequence(input, lengths, batch_first=False):
+    return g.op("PackPadded", input, lengths, outputs=2)
+
+
+def pack_padded_sequence(input, *args, **kwargs):
```
I don't think you need to do this additional invocation - just apply `@symbolic_override_first_arg_based(_ _symbolic_pack_padded_sequence)` to the function below (see https://github.com/pytorch/pytorch/blob/5b43c22f73279c67084a2357a489420c705cb84f/torch/onnx/__init__.py#L381)

The reason why we were arguing about perf for the main RNN function was that the internal function `forward` was declared (and decorated) on each RNN() function invocation. In this case the decorator would run only once on module loading so the resulting code from decorator would be equivalent to what you wrote here.

anderspapitto replied:
that would be great, but it doesn't work because importing torch.onnx at the top level is a circular dependency

ezyang replied:
You've lost the public-facing docblock for `pack_padded_sequence` here. Move it from the internal function to this one.

anderspapitto replied:
++

dzhulgakov replied:
argh, circular dependency is BS. We can extract those decorators from the onnx submodule and put them somewhere else. It's really ugly to proliferate these kind of hacks around. @ezyang - what's the regular practice for circular deps in pytorch? Was there some design of which submodules can depend on others?

anderspapitto replied:
I'd rather fix this as a separate issue; this diff doesn't lock us into doing it this way if someone finds a better way down the line

ezyang replied:
There isn't a circular dependency discipline at the moment; but I would definitely not oppose moving these combinators to another module. Our current practice is, however, to push onnx imports inside functions precisely to avoid this problem.

dzhulgakov commented on torch/nn/utils/rnn.py:
```diff
@@ -158,7 +176,21 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0):
+    return g.op("PadPacked", input.data, input.batch_sizes, outputs=2)
+
+
+def pad_packed_sequence(input, *args, **kwargs):
```
same as above

dzhulgakov commented on torch/onnx/symbolic.py:
```diff
@@ -554,16 +555,12 @@ def Elman_RNN_symbolic_builder(
         prev_output = input
         h_outs = []
+
+        sequence_lens = g.op("Undefined") if batch_sizes is None else batch_sizes
```
nit: maybe declare `g.op(Undefined)` as 'unused' somewhere to make it look nicer?

dzhulgakov commented on torch/onnx/symbolic.py:
```diff
@@ -579,29 +576,25 @@ def symbolic(g, input, all_weights, h0, **fkwargs):
-            return _unimplemented("LSTM", "dropout")
-        if bidirectional:
-            return _unimplemented("LSTM", "bidirectional")
+    if batch_first:
```
move it back to `symbolic`. In my understanding we want to throw inside `symbolic` to get slightly better error messages attributed to a particular node

anderspapitto replied:
++, I think this code got changed underneath me

dzhulgakov replied:
Yep, I changed it :)

dzhulgakov commented on torch/onnx/symbolic.py:
```diff
@@ -623,23 +616,20 @@ def symbolic(g, input, all_weights, h0_and_c0, **fkwargs):
-        # optional arguments.
-        sequence_len, batch_size = input.type().sizes()[0:2]
-        sequence_lens = g.op('Constant', value_t=torch.IntTensor(batch_size).fill_(sequence_len))
+    if batch_first:
```
same here

dzhulgakov reviewed (commented):
So what are the breaking changes for user? Need to wrap lengths into a tensor?

@soumith - do we know for existing public repos how much it'd affect them? One option might be to still accept both but complain in symbolic export if non-tensor is passed (though it's hacky).

Otherwise - looks great, thanks for diligently working on it. Please find some minor notes inline about final polishing.

anderspapitto commented:
@pytorchbot retest this please

anderspapitto commented:
@dzhulgakov yeah, need to wrap list of integers with Variable()

Although, actually I expect that in many cases, people will already have the lengths stored in a Variable, and were having to explicitly convert to list of int (because semantically, lengths/batch sizes are just as much a feature of the dynamic input as the actual float values) - so actually they will just be able to remove any such code.

apaszke commented:
@anderspapitto I don't think that's the case. `PackedSequence` object are pretty much only produced using functions that we provide, so I doubt anyone has those lengths packed as Variables upfront.

jekbradbury commented:
The `batch_sizes` attribute is internal to `PackedSequence` objects, but `lengths` needs to be provided by the user (and providing it as a `Variable`) is more natural.

anderspapitto commented:
@apaszke I actually don't have many examples, but it is true in the one (unfortunately internal, so I can't link it) case that I have looked at.

Anyway I think I didn't say that in the clearest way, so let me give it another go:

Whenever a user calls `PackPadded(input, lengths)`, I would assume that `input` and `lengths` almost always came from the same source. Therefore it should probably be very natural to turn `lengths` into a Variable at the same time that `input` is turned into a Variable (which already happens).

apaszke commented on torch/autograd/function.py:
```diff
@@ -288,15 +288,22 @@ def _iter(obj):
+                res_e, input = unflatten_helper(input, e)
+                res.append(res_e)
+        if isinstance(proto, PackedSequence):
+            return PackedSequence(*res), input
```
Wouldn't changing the code to `type(proto)(*res)` be enough? `isinstance` is very expensive

apaszke replied:
Ah sorry, this will fail for regular tuples, because they expect a single argument in their constructor.

apaszke replied:
Actually where's this even called? I can't see it in the diff

ezyang replied:
@anderspapitto and I agreed that we'll do `type(proto)(res)` and add a constructor to `PackedSequence` which takes a regular tuple.

anderspapitto replied:
no, because while an n-tuple can be constructed from another n-tuple `x` by saying `tuple(x)`, an n-namedtuple cannot, it must be constructed from n individual arguments.

I'll address this by changing the constructor of PackedSequence() to also accept a single tuple

apaszke replied:
We can change the constructor to always match that of `tuple`. There's a note in the docs that says that it should never be instantiated by the user, but only constructed using the API for packed sequences

dzhulgakov replied:
See my comment above - `namedtuple` has `_make` method which takes the iterable. So you can do `if hasattr(type(proto), '_make')` which is hopefully faster than isinstance

anderspapitto replied:
I went with changing the constructor to allow either style of arg-passing

ezyang commented on test/test_nn.py:
```diff
@@ -2958,9 +2958,11 @@ def compare_cpu_gpu(outputs_cpu, outputs_gpu):
-                    batch_sizes = [7, 5, 5, 2, 1, 1]
-                    input_val = rnn_utils.pack_padded_sequence(input_val, batch_sizes, batch_first=batch_first)
-                    grad_output = rnn_utils.pack_padded_sequence(grad_output, batch_sizes, batch_first=batch_first).data
+                    lengths = torch.IntTensor([7, 5, 5, 2, 1, 1])
```
The old code should continue to work even if we pass it a list directly, otherwise it's BC-breaking, right? Can we have a warning for when this use case happens?

ezyang replied:
Concretely:

1. If some code exercises the externally facing API, as this code does, you should keep the test for the legacy API because someone might be using it, and we need to continue testing that it works.
2. You should add a warning to `pack_padded_sequence` that detects if you get a Python list and throws a deprecation notice.

anderspapitto replied:
it is BC-breaking. I'll add the shim you've asked for below.

ezyang commented on torch/nn/_functions/packing.py:
```diff
@@ -14,7 +14,7 @@ def forward(ctx, input, lengths, batch_first):
         steps = []
         batch_sizes = []
-        lengths_iter = reversed(lengths)
+        lengths_iter = reversed(list(lengths))
```
Because `lengths` used to be a list but now it is not, it is worth commenting that `lengths` is a `Tensor` here.

anderspapitto replied:
++

ezyang commented on torch/nn/_functions/rnn.py:
```diff
@@ -70,7 +70,7 @@ def StackedRNN(inners, num_layers, lstm=False, dropout=0, train=True):
     total_layers = num_layers * num_directions
 
-    def forward(input, hidden, weight):
+    def forward(input, hidden, weight, batch_sizes):
```
@apaszke We need to confirm that these functions are not part of the public facing API. They're all defined and used solely in an underscore module, so I think this is the case.

apaszke replied:
They're not, but why do we need to change the way we thread batch_sizes through those functions? This should be invisible to the tracer anyway

anderspapitto replied:
they need to be visible to the tracer. How can the exported ONNX model run if it doesn't have access to lengths/batch_sizes? It's the same point as I made above - semantically, `(input, lengths)` is actually a single input.

ezyang commented on torch/nn/utils/rnn.py:
```diff
@@ -130,19 +147,20 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0):
+        Tuple of Variable containing the padded sequence, and Variable
+        containing the list of lengths of each sequence in the batch.
+
     """
```
Here is where the BC code if a user feeds us a `list` as `batch_sizes` should live; we should generate a Variable LongTensor on the fly in that case.

anderspapitto replied:
I had some backwards-compat code that worked a little differently earlier, and then was asked to take it out (I previously had two attributes, batch_size_as_list and batch_size_as_variable). I'll add back the version you're asking for here

ezyang commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -167,6 +172,84 @@ void fuseTransposeIntoGemm(std::shared_ptr<Graph>& graph) {
+
+    // remove PackPadded from in front of the RNN
+    n->outputs()[0]->replaceFirstUseWith(n->inputs()[0]);
+    n->outputs()[1]->replaceFirstUseWith(n->inputs()[1]);
```
Should there be an ASSERT that `n->outputs()[1]` is used only once?

anderspapitto replied:
no, the assert would fire. see my reply above

ezyang commented on torch/nn/modules/rnn.py:
```diff
@@ -118,7 +118,7 @@ def check_forward_args(self, input, hidden, batch_sizes):
 
         if is_input_packed:
-            mini_batch = batch_sizes[0]
+            mini_batch = batch_sizes.data[0]
```
See below.

ezyang commented on torch/nn/utils/rnn.py:
```diff
@@ -130,19 +147,20 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0):
     """
     var_data, batch_sizes = sequence
-    max_batch_size = batch_sizes[0]
+    max_batch_size = list(batch_sizes.data)[0]
```
I think `int(batch_sizes[0])` is more idiomatic here.

ezyang commented on torch/nn/utils/rnn.py:
```diff
@@ -130,19 +147,20 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0):
     prev_batch_size = batch_sizes[0]
     prev_i = 0
-    for i, batch_size in enumerate(batch_sizes + [0]):
+    for i, batch_size in enumerate(list(batch_sizes.data) + [0]):
```
More idiomatic: `list(batch_sizes.data)` => `batch_sizes.tolist()`

anderspapitto replied:
++

ezyang commented on torch/nn/utils/rnn.py:
```diff
@@ -158,7 +176,21 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0):
     if batch_first:
         output = output.transpose(0, 1)
-    return output, lengths
+    return output, Variable(torch.LongTensor(lengths))
```
This needs a comment saying that this Variable doesn't actually have any history, it is purely here to make ONNX export easier.

anderspapitto replied:
++

anderspapitto commented:
addressed all comments

anderspapitto commented:
```
23:37:49 + python ../compare_with_baseline.py test_gpu_speed_word_language_model 115.893 116.011 115.957 116.066 116.069
23:37:49 baseline mean:  115.5332
23:37:49 baseline sigma:  0.10897
23:37:49 test mean:  115.9992
23:37:49 test sigma:  0.06713091687144009
23:37:49 z-value:  4.2764063503717376
23:37:49 Traceback (most recent call last):
23:37:49   File "../compare_with_baseline.py", line 46, in <module>
23:37:49     raise Exception("z-value >= 3, there is 99.7% chance of perf regression.")
23:37:49 Exception: z-value >= 3, there is 99.7% chance of perf regression.
```

how can I know if this is spurious or something I need to address? 116.0 vs 115.5 seems small but maybe it isn't?
